### DISCUSSION TITLE
Dotnet support for FreeBSD

### DISCUSSION THREAD

**Discussion by lillo42** [dotnet/runtime#102377] [Q&A] [answered]
Hey folks, I'm searching for dotnet on FreeBSD, I found some issues related to that but something isn't clear to me what is missing for dotnet to be completely compatible with FreeBSD (and when a version is released is to be released for FreeBSD too)

**Comment by neon-sunset** [comment]
There are no Microsoft-maintained FreeBSD builds. There is, now, however https://www.freshports.org/lang/dotnet/

In theory, it should be accessible with just `pkg install lang/dotnet`.

**Comment by am11** [comment]
FreeBSD is a community supported platform in .NET. You can install it from ports which @neon-sunset pointed out. It uses the same source-build approach as linux distros (https://github.com/dotnet/dotnet). If you run into a problem, like for any platform, open an issue in runtime repo (existing issues: https://github.com/dotnet/runtime/issues?q=is:issue+label:os-freebsd).

>  what is missing for dotnet to be completely compatible with FreeBSD

@Thefrank may have a better answer, but it is usable as-is and has been for a while. Last time I tried compiling basic apps, dotnet(1) was working fine on FreeBSD host as well as dotnet(1) for linux was working under Linuxulator (by disabling W^X). Haven't tried in a while so YMMV.

**Comment by Thefrank** [comment, marked as answer]
> what is missing for dotnet to be completely compatible with FreeBSD

Depending on how you see "completely compatible" either nothing at all or so many interlocking things that this topic would quickly expand far beyond the question.

As @am11 said the `dotnet` in FreeBSD's ports is usable right now for basic needs. You will need to turn off the ASLR flag for any binaries you generate with it (`elfctl -e NAME`). If you want to do serious development with it for/on FreeBSD and want to use anything more than the SDK (e.g., using `dotnet tool ...`) you will run into typical THING-is-supported-by-the-community limitations as developers are hesitant to build for/support a platform that is not officially supported by the language they are using.

@lillo42 I hope that clears things up without getting too complicated!

**Reply by am11** [reply]
@Thefrank, thanks!

Is there an issue tracking ASLR / `elfctl -e NAME`, and does it apply to AOT published binaries?

We had a similar issue with linux binaries during the debugging (in diagnostics repo) a while back and I opened https://github.com/llvm/llvm-project/issues/61899. Workaround was clear, but I didn't knew FreeBSD had it for even non-debug run. Something we could fix in apphost (assuming AOT binaries run fine without disabling ASLR).

**Reply by am11** [reply]
Just tested in CirrusCI terminal access, FreeBSD 14 amd64:

```sh
$ uname -a
FreeBSD cirrus-task-5661843576848384 14.0-RELEASE FreeBSD 14.0-RELEASE #0 releng/14.0-n265380-f9716eee8ab4: Fri Nov 10 05:57:23 UTC 2023     root@releng1.nyi.freebsd.org:/usr/obj/usr/src/amd64.amd64/sys/GENERIC amd64

$ pkg install -y dotnet-8.0.0

$ dotnet new webapi -n w2
$ cd w2
$ dotnet publish -o dist -p:PublishSingleFile=true

$ ls dist
appsettings.Development.json    w2
appsettings.json                w2.pdb

$ file dist/w2
dist/w2: ELF 64-bit LSB pie executable, x86-64, version 1 (FreeBSD), dynamically linked, interpreter /libexec/ld-elf.so.1, for FreeBSD 14.0 (1400097), FreeBSD-style, BuildID[sha1]=2cb4a216e1fa2ca413fe0119fd0a092ccf7d44d5, stripped

# check if ASLR is enabled on the system (yes it is)
$ sysctl kern.elf64.aslr.enable
kern.elf64.aslr.enable: 1

# check if binary has ASLR disabled (no it is not)
$ elfctl dist/w2
File 'dist/w2' features:
noaslr          'Disable ASLR' is unset.
noprotmax       'Disable implicit PROT_MAX' is unset.
nostackgap      'Disable stack gap' is unset.
wxneeded        'Requires W+X mappings' is unset.
la48            'amd64: Limit user VA to 48bit' is unset.

$ dist/w2
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /tmp/cirrus-ci-build/w2
```

Seems to be running with ASLR based on this simple test.  👀

**Comment by Thefrank** [comment]
EDIT: This should've been a reply to @am11 above but it turns out I can't click the right button to post replies correctly

https://github.com/freebsd/freebsd-ports/blob/698eadbe278c0e8c50da6082994e16bb4910a398/lang/dotnet/Makefile#L99
was added because of issues brought up https://github.com/dotnet/runtime/issues/14537#issuecomment-1510996537 and reproduced https://github.com/dotnet/runtime/issues/14537#issuecomment-1515938037

Looping the script found in the second comment, I have not been able to reproduce the issue on FreeBSD 14. For libunwind 1.8 I still had ASLR off system-wide on my build system as I was sometimes hitting crashes. The current libunwind on FreeBSD is 20240221 which is 1.8.1 and the one used for the test just now. I try something more complex with AZP and some dotnet builds a bit later and edit this with an update

EDIT: neither the newest versions of jellyfin nor azure-pipelines-agents experienced crashes during the building or testing process. Maybe it was `libunwind` after all? 🤷

**Reply by am11** [reply]
Cool! Your clear infos always bring delight to the conversation, regardless of the thread. 👌

Regarding runtime packs (apphost, crossgen2 and the upcoming NativeAOT ones), are they usually 1) published to nuget.org, 2) packed with the sdk, 3) delivered as a sub/separate package in ports, 4) part of the dotnet-X.Y main package itself or 5) a separate nuget feed hosted by the FreeBSD org? I _think_ Ubuntu and Alpine are using approach 3, not sure what FreeBSD is using.

Looking forward to get hands on NativeAOT runtime pack in .NET 9. 👍

(ps: asking that question because when nativeaot / ilc runtime pack installation hits the network in each run, it degrades the CI experience a bit.. which is the case with Microsoft builds.. so I use CI/docker caching which varies by system; CirrusCI, GH Actions etc.)

**Reply by Thefrank** [reply]
> Regarding runtime packs (apphost, crossgen2 and the upcoming NativeAOT ones), are they usually 1) published to nuget.org, 2) packed with the sdk, 3) delivered as a sub/separate package in ports, 4) part of the dotnet-X.Y main package itself or 5) a separate nuget feed hosted by the FreeBSD org? I _think_ Ubuntu and Alpine are using approach 3, not sure what FreeBSD is using.
 
I would like nuget.org (1) to have them as that is the default location that is used for most products. However I think that would only be possible if Microsoft publish the NuGets. (2) I think only NET9+ supports that via `BundleRuntimePacks` + `BundleNativeAotCompiler`? (3) with source building this might be the easiest option and should not take much more work to add as a build option (4) also possible like (3); I personally favor (3) but have no issues with (4) other than preference. (5) This would be up to the foundation and core team and is less easy to access than the other options however I will take anything at this point :)

> Looking forward to get hands on NativeAOT runtime pack in .NET 9. 👍

.NET8 should support NativeAOT already? Or did you mean having the pack included without having to source it externally?
```
root@freebsd:~/test # uname -a
FreeBSD freebsd 13.3-RELEASE FreeBSD 13.3-RELEASE releng/13.3-n257428-80d2b634ddf0 GENERIC amd64

root@freebsd:~/test # dotnet new console
The template "Console App" was created successfully.

Processing post-creation actions...
Restoring /root/test/test.csproj:
  Determining projects to restore...
  Restored /root/test/test.csproj (in 137 ms).
Restore succeeded.


root@freebsd:~/test # dotnet publish -r freebsd-x64 -o out/ /p:PublishAot=true
MSBuild version 17.8.3+195e7f5a3 for .NET
  Determining projects to restore...
  Restored /root/test/test.csproj (in 259 ms).
  test -> /root/test/bin/Release/net8.0/freebsd-x64/test.dll
  Generating native code
  test -> /root/test/out/

root@freebsd:~/test # file out/test
out/test: ELF 64-bit LSB pie executable, x86-64, version 1 (FreeBSD), dynamically linked, interpreter /libexec/ld-elf.so.1, FreeBSD-style, BuildID[sha1]=4188d02da79b60a76f63a4e2d84a2f7421396813, stripped

root@freebsd:~/test # ./out/test
Hello, World!
```
`webapi` produces a few IL3053 and IL2104 warnings but finishes and runs without issue

**Reply by am11** [reply]
> I would like nuget.org (1) to have them as that is the default location that is used for most products. However I think that would only be possible if Microsoft publish the NuGets.

I think anyone can upload a runtime package by that name. There is no restriction (per my understanding of what @agocke once explained).

> (2) I think only NET9+ supports that via BundleRuntimePacks + BundleNativeAotCompiler? 

The only thing I know is from @akoeplinger's explaination https://github.com/dotnet/sdk/pull/37872#issuecomment-1889514656 and that makes the CI scenario (which needs to always run from scratch) slower specifically for PublishAot (half of 1/3 comprise of stuff like debug symbols which 99.9999% of the people never use in real-world https://github.com/dotnet/runtime/issues/102684), unless we setup the per-system caching...  + nugets don't seem to hit network cache, so "downloading ilc package" is the longest time spent on publishing microservice in my workplace in stock containers (unless we modify them with caches and whatnot). :(

> (3) with source building this might be the easiest option and should not take much more work to add as a build option

Yup, 3 or 4 sounds great! One installation step (high probability of hitting the cache) and done; rest of the time spent on building user app.

> .NET8 should support NativeAOT already? Or did you mean having the pack included without having to source it externally?

Ah on CirrusCI, `pkg install -y dotnet-8.0.0` followed by `dotnet publish -p:PublishAot=true` is giving me:
```
error NU1101: Unable to find package runtime.freebsd-x64.Microsoft.DotNet.ILCompiler.
No packages exist with this id in source(s): nuget.org
```

so i thought we have to wait for .NET 9 (for source builds to provide ilc packages).

**Reply by akoeplinger** [reply]
> I think anyone can upload a runtime package by that name. There is no restriction [...]

The runtime packs like https://www.nuget.org/packages/Microsoft.NETCore.App.Runtime.NativeAOT.linux-x64 or https://www.nuget.org/packages/runtime.linux-x64.Microsoft.DotNet.ILCompiler are using reserved prefixes on nuget.org so no one else can upload these packages.

There was a policy decision that we will only push to nuget.org packages for platforms that MS officially supports so community-supported platforms or third-party supported platforms like s390x need to put them somewhere else, e.g. in a custom nuget feed or somewhere else the SDK looks in. See also this thread: https://github.com/dotnet/source-build/issues/1215

**Reply by am11** [reply]
Great! Just read https://learn.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation. So `packs` sound like a good place for them to be (i.e. included in the dotnet, dotnet-sdk or similar distro package). Per distro nuget feed is probably another cleaner option, whose feed can be included in distros's SDK in `~/.nuget/NuGet/NuGet.Config` global config, but it comes with additional maintenance overhead.

**Reply by Thefrank** [reply]
> so i thought we have to wait for .NET 9 (for source builds to provide ilc packages).

On the FreeBSD [this](https://reviews.freebsd.org/D44561) was proposed to modify the `dotnet` port to add NuGets to their own location and register it with dotnet. It should be sufficient for .NET8 until .NET9 where it can be re-approached (bundled back package vs bundled by SDK)

It should cover all the required NuGets:
```
%%NUPKG%%dotnet/packages/Microsoft.AspNetCore.App.Runtime.freebsd-x64.8.0.0.nupkg
%%NUPKG%%dotnet/packages/Microsoft.NETCore.App.Crossgen2.freebsd-x64.8.0.0.nupkg
%%NUPKG%%dotnet/packages/Microsoft.NETCore.App.Host.freebsd-x64.8.0.0.nupkg
%%NUPKG%%dotnet/packages/Microsoft.NETCore.App.Runtime.freebsd-x64.8.0.0.nupkg
%%NUPKG%%dotnet/packages/runtime.freebsd-x64.Microsoft.DotNet.ILCompiler.8.0.0.nupkg
%%NUPKG%%dotnet/packages/runtime.freebsd-x64.Microsoft.NETCore.DotNetAppHost.8.0.0.nupkg
%%NUPKG%%dotnet/packages/runtime.freebsd-x64.Microsoft.NETCore.DotNetHost.8.0.0.nupkg
%%NUPKG%%dotnet/packages/runtime.freebsd-x64.Microsoft.NETCore.DotNetHostPolicy.8.0.0.nupkg
%%NUPKG%%dotnet/packages/runtime.freebsd-x64.Microsoft.NETCore.DotNetHostResolver.8.0.0.nupkg
%%NUPKG%%dotnet/packages/runtime.freebsd-x64.Microsoft.NETCore.ILAsm.8.0.0.nupkg
%%NUPKG%%dotnet/packages/runtime.freebsd-x64.Microsoft.NETCore.ILDAsm.8.0.0.nupkg
%%NUPKG%%dotnet/packages/runtime.freebsd-x64.Microsoft.NETCore.TestHost.8.0.0-rtm.23531.3.nupkg
```

Projects that come with their own `nuget.config` will still need to be modified manually to include the FreeBSD specific NuGet location but I can not see that as being much of an issue.

**Reply by am11** [reply]
Yup, if they are using `<clear/>` statement in their feed, then it won't inherit the global.json. If they restored a project once before the restoring using global nuget config, it will hit the local cache, but it's a clunky workflow I agree. 😅

Regarding https://reviews.freebsd.org/D44561, is the current behavior of cirrus-ci (unable to find ILCompiler package) expected (i.e. we have to wait for next quarter)? I will try to fast track from quarterly feed next.

**Reply by am11** [reply]
Upgrading from quarterly to latest didn't help on cirrus-ci:

```sh
$ sed -i '' -e '/^\[quarterly\]/,/^$/ s/^url:/#url:/' -e '/^\[latest\]/,/^$/ s/^#url:/url:/' /etc/pkg/FreeBSD.conf
$ pkg update && pkg upgrade -y && pkg install -y dotnet-8.0.0
$ dotnet new webapiaot -n w1
dotnet new webapiaot -n w1
...
The template "ASP.NET Core Web API (native AOT)" was created successfully.

Processing post-creation actions...
Restoring /tmp/cirrus-ci-build/w1/w1.csproj:
  Determining projects to restore...
/tmp/cirrus-ci-build/w1/w1.csproj : error NU1101: Unable to find package runtime.freebsd-x64.Microsoft.DotNet.ILCompiler. No packages exist with this id in source(s): nuget.org
  Failed to restore /tmp/cirrus-ci-build/w1/w1.csproj (in 688 ms).
Restore failed.
Post action failed.
Manual instructions: Run 'dotnet restore'
```
Perhaps the post-install step isn't running?

**Reply by Thefrank** [reply]
> Regarding https://reviews.freebsd.org/D44561, is the current behavior of cirrus-ci (unable to find ILCompiler package) expected (i.e. we have to wait for next quarter)?

Yes. It is still waiting to be approved/commited.

Once commited you will have to wait until next quarter unless you want to switch to using `latest` package mirror. `latest` will pick up the new package once it reaches pkg mirrors (~1-4 days after commit). `quarterly` is the default and while `latest` will be just as reliable I like to keep local build systems on `quarterly` unless I absolutely need something faster due to demands (e.g., the rare rust project that require the absolute newest everything). Both `quarterly` and `latest` receive security updates for packages.

**Reply by am11** [reply]
Ah, didn't noticed "Needs Review" status. 😅

Yup, true for local dev environment, but the CirrusCI instances are of a throw-away sort; we can modify it however we see fit for the job, once it exists, it gets disposed. So `latest` in that sandbox-y environment is fine (at least I'm not building anything production-ready there yet).

**Reply by ghost** [reply]
Does it support native aot now? Some reference packages in  `/usr/local/share/dotnet/library-packs`. But when I using `dotnet publish -p:PublishAot=true` it still throw error. Anyone knows?

**Reply by Thefrank** [reply]
@Jozhden I just tested and this works fine for net10-RC1. This should also work for net9. This might not be the case for net8.

**Reply by ghost** [reply]
> @Jozhden I just tested and this works fine for net10-RC1. This should also work for net9. This might not be the case for net8.

Thanks . But I am following the best practice and install it by `pkg install dotnet` Then I am trying  a simple test `dotnet new console --aot` . It seems not working in this case which is .net9. Is there any thing I can do ?

**Reply by Thefrank** [reply]
@Jozhden I assume you are on `latest` and not `quarterly` for packages? 
Is this your error? `error NU1101: Unable to find package runtime.freebsd-x64.Microsoft.DotNet.ILCompiler. No packages exist with this id in source(s): /usr/local/share/dotnet/library-packs, nuget.org`. If yes, then it appears Restore is looking for the portable version and not the non-portable version that is packaged in `library-packs` (`runtime.freebsd.14-x64.Microsoft.DotNet.ILCompiler.9.0.6.nupkg`). This issue was resolved in a later SDK version (9.0.2xx) but VMR (and thus FreeBSD ports) does support that branch (9.0.1xx only). This hacky workaround might also work for ILC https://github.com/dotnet/runtime/issues/111900#issuecomment-2732128285
